torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+        for_loop->iter_domain(),
+        for_loop->index(),
+        for_loop->start(),
+        for_loop->stop(),
+        for_loop->step(),
+        for_loop->vectorize(),
+        for_loop->vectorize_shift());
+    for (auto expr : new_loop_body) {
+      new_loop->body().push_back(expr);
+    }
+    return new_loop;
+  }
+
+  kir::IfThenElse* eliminateDeadCode(kir::IfThenElse* ite) {
+    auto new_then_body = eliminateDeadCodeInScope(ite->thenBody().exprs());
+    auto new_else_body = eliminateDeadCodeInScope(ite->elseBody().exprs());
+    if (new_then_body.empty() && new_else_body.empty()) {
+      return nullptr;
+    }
+    auto new_ite = ir_builder_.create<kir::IfThenElse>(ite->predicate());
+
+    for (auto expr : new_then_body) {
+      new_ite->thenBody().push_back(expr);
+    }
+    for (auto expr : new_else_body) {
+      new_ite->elseBody().push_back(expr);
+    }
+    return new_ite;
+  }
+
+ private:
+  std::unordered_set<kir::TensorView*> live_tvs_;
+  std::unordered_set<kir::TensorView*> dead_tvs_;
+  std::vector<kir::TensorView*> candidate_tvs_;
+  std::unordered_set<kir::TensorView*> candidate_tv_set_;
+
+  std::vector<kir::Expr*> result_exprs_;
+  kir::IrBuilder ir_builder_;
+};
+
+//! A pass to eliminate redundant parallel broadcasts that are consumers
+//!  of warp reduction.
+//! Detects the following pattern:
+//!
+//!  For ... (serial)
+//!   For ... (serial)
+//!     T1[0] = warp_reduce (T0[0])
+//!     T2[0] = block_broadcast (T1[0])
+//!
+//!  The block_broadcast can then be eliminated given that both the warp
+//!   reduce and the broadcast are known in compile-time to be parallelized
+//!   on a single warp only.
+//!
+//!  Currently only limited to buffers of size-1 to avoid having to
+//!   re-run indexing
+//!
+//!  This pass operates in 3 phases:
+//!   1. fuseBroadcastWithWarpReduce identifies the broadcasts that can
+//!    be removed, and generates a replacement map from the broadcast
+//!    output to reduction output.
+//!
+//!   2. kir_utils::replaceInputsInExpr replaces applicable uses of
+//!    the broadcast output with the corresponding reduction output.
+//!
+//!   3. eliminateDeadBroadcastAndAllocate removes the broadcast ops
+//!    and corresponding allocations if they're un-used after step 2.
+class fuseBroadcastWithWarpReduce {
+ public:
+  static std::vector<kir::Expr*> fuse(const std::vector<kir::Expr*>& exprs) {
+    fuseBroadcastWithWarpReduce fuse_broadcast_map(exprs);
+    const auto replaced_inputs = kir_utils::replaceInputsInExpr(
+        exprs, fuse_broadcast_map.val_replacement_map_);
+    return eliminateDeadBroadcastAndAllocate::run(replaced_inputs);
+  }
+
+ private:
+  fuseBroadcastWithWarpReduce(const std::vector<kir::Expr*>& exprs) {
+    // open stack space for global scope
+    running_kir_tv_to_allocate_map_.emplace_back(
+        std::make_unique<
+            std::unordered_map<kir::TensorView*, kir::Allocate*>>());
+    running_visible_allocation_stack_.emplace_back(
```

Comment:
Does it really need to manage allocation scopes using a stack? When KIR is traversed, allocation nodes should be visited before expressions defining and using allocated tensors. Since each tensor is allocated and defined once and uniquely, a simple flat map seems sufficient. Or, maybe I'm missing something around aliasing?

Reply:
Good quesiton. I guess it is not needed if each tensor is allocated once and uniquely. This pass isn't assuming that and I'm not sure if we want to assume that here since lower_allocation could also be extended/optimized long term.

Another reason for the stack is for making sure the broadcast and reduce are in the same loopnest, and producing/consuming the same local buffer. This was also trying to make minimal assumption about the previous passes.

Reply:
OK, makes sense. Maybe adding your response as a comment in the code would be helpful.